modules/correlation/group-lines.c

```diff
@@ -0,0 +1,300 @@
+                    evt_tag_str("key", context->super.key.session_id),
+                    evt_tag_str("line", context->line_buffer->str));
+          return GP_CONTEXT_COMPLETE;
+        }
+      else if (verdict & MLL_REWIND_SEGMENT)
+        {
+          msg_debug("group-lines: accumulated line extracted",
+                    evt_tag_str("key", context->super.key.session_id),
+                    evt_tag_str("line", context->line_buffer->str));
+          return GP_STARTS_NEW_CONTEXT;
+        }
+      else
+        g_assert_not_reached();
+    }
+  else if (verdict & MLL_WAITING)
+    {
+      if (verdict & MLL_CONSUME_SEGMENT)
+        {
+          _update_context_add_message(self, context, msg, line, line_len);
+          return GP_CONTEXT_UPDATED;
+        }
+      else
+        g_assert_not_reached();
+      return FALSE;
+    }
+  else
+    g_assert_not_reached();
+
+}
+
+static LogMessage *
+_generate_synthetic_msg(GroupLines *self, GroupLinesContext *context)
+{
+  LogPathOptions path_options = LOG_PATH_OPTIONS_INIT;
+
+  path_options.ack_needed = FALSE;
+
+  LogMessage *msg = log_msg_ref(correlation_context_get_last_message(&context->super));
+  log_msg_make_writable(&msg, &path_options);
+  return msg;
+}
+
+static LogMessage *
+_aggregate_context(GroupingParser *s, CorrelationContext *c)
+{
+  GroupLines *self = (GroupLines *) s;
+  GroupLinesContext *context = (GroupLinesContext *) c;
+
+  LogMessage *msg = _generate_synthetic_msg(self, context);
+
+  log_msg_set_value(msg, LM_V_MESSAGE, context->line_buffer->str, context->line_buffer->len);
+
+  return msg;
+}
+
+static const gchar *
+_format_persist_name(const LogPipe *s)
+{
+  static gchar persist_name[512];
+  GroupLines *self = (GroupLines *)s;
+
+  g_snprintf(persist_name, sizeof(persist_name), "group-lines(%s,scope=%d,clone=%d)", self->super.key_template->template,
+             self->super.scope, self->clone_id);
+  return persist_name;
+}
+
+static gboolean
+_init(LogPipe *s)
+{
+  GroupLines *self = (GroupLines *) s;
+
+  if (!multi_line_options_validate(&self->multi_line_options))
+    return FALSE;
+  return grouping_parser_init_method(s);
+}
+
+static LogPipe *
+_clone(LogPipe *s)
+{
+  GroupLines *self = (GroupLines *) s;
+  GroupLines *cloned;
+
+  cloned = (GroupLines *) group_lines_new(s->cfg);
+
+  log_parser_set_template(&cloned->super.super.super, self->super.super.super.template);
+  grouping_parser_set_key_template(&cloned->super.super.super, self->super.key_template);
+  grouping_parser_set_sort_key_template(&cloned->super.super.super, self->super.sort_key_template);
+  grouping_parser_set_timeout(&cloned->super.super.super, self->super.timeout);
+  grouping_parser_set_scope(&cloned->super.super.super, self->super.scope);
+  group_lines_set_separator(&cloned->super.super.super, self->separator);
+
+  multi_line_options_copy(&cloned->multi_line_options, &self->multi_line_options);
+
+  cloned->clone_id = self->clone_id++;
```

Comment:
[optional]

It seems weird to me, that we change the `clone_id` of `self` here. Same applies for `groupingby`.

Reply:
this is because we need to differentiate all clones as they will all use their own state.

Reply:
Yes, I agree, but I think we should not modify `self` with `self->clone_id++` in the clone fn. IMO `cloned->clone_id = self->clone_id + 1` would be better.

Reply:
the issue is there might be multiple clones and each needs a different ID.

Reply:
Thanks for the explanation, I understand now why it is done like this. I don't think we should change it in this PR, as groupingby already has this logic, but I think this is not the way to go in the long run. My problem is: The persist name of the parser is dependent of how many times it was cloned. Let's say we have a config, where it gets cloned 2 times. It has `clone_id=2` in its persist name and stores its values in the persist with that name. We add another reference/user of the parser in the config, and reload. Now it gets `clone_id=3`, and it won't find its persist entry from before reload and starts a new, however the new clone gets an already filled persist entry.